### PR TITLE
Try drastically simplifying travis

### PR DESCRIPTION
**Pull request by jchodera**
Just experimenting for now.

### CODE REVIEW

**Comment by jchodera** [comment]
@pgrinaway @juliebehr : I think this is probably the way to go. It's a lot simpler, ensures we won't accidentally miss running new tests if we forget to update `.travis.yml` when we add a new test, and takes less time (29 minutes) than the existing parallelized travis ([34 min 50 sec according to the last good run](https://travis-ci.org/choderalab/perses/builds/174349214)).

**Comment by jchodera** [comment]
Ready to review and merge.

**Comment by jchodera** [comment]
@pgrinaway @juliebehr Please review and merge.

**Comment by jchodera** [comment]
Thanks!

**Comment by pgrinaway** [comment]
Looks like the protonation state test system is broken.

**Comment by pgrinaway** [comment]
Actually, I think the mutation proposals are broken too.

**Comment by pgrinaway** [comment]
Maybe we can punt on those, since they aren't essential for the first paper.

**Comment by jchodera** [comment]
We probably missed testing these because of the split travis.

I'll fix the protonation state stuff; @juliebehr can probably easily fix the mutation tests, or else we can punt if it looks like it will take more than a few min.

**Comment by jchodera** [comment]
Can you override and merge? The current [master build is broken anyway](https://travis-ci.org/choderalab/perses/builds/175942355), so there's no point in requiring the build pass.

**Comment by pgrinaway** [comment]
>Can you override and merge? The current master build is broken anyway, so there's no point in requiring the build pass.

I don't think that's a good justification, since there are many reasons a build can fail. We should figure out how master is broken--PRs must pass to be merged, so I assume a dependency changed?

**Comment by pgrinaway** [comment]
I wonder if some change in openmoltools caused this.

**Comment by jchodera** [comment]
I don't know what you're suggesting here. 

Yes, we should figure out what caused `master` to fail, but what do you propose to do about this PR? Implement all the fixes to protonation states and mutations into this PR and then merge it? Or somehow separately implement the fixes to protonation states and mutations in separate PRs, along with their additional test coverage, merge those, and then merge the refactoring of tests? That sounds like a colossal waste of time.

The major problem here is that we discovered we have incomplete test coverage, and now we have to fix that. There are probably lots of things to fix. So we should probably merge the tests, then fix the broken things.

**Comment by pgrinaway** [comment]
I'm suggesting we:

1. Fix master -- the fact that it is broken despite every recent PR being "build passing" is concerning. We should figure out what's wrong here.

2. Fix this PR--perhaps excluding the mutation tests and fixing the imidazole stuff here until this PR passes.

3. Merge this PR.


Convolving the failures here with the ones in master seems like a way to make debugging harder.

**Comment by pgrinaway** [comment]
>There was just an upstream OpenMM change to how bonds are handled in OpenMM. This issue looks like it might be related:

Is there some way to not always use the latest code? It strikes me that this guarantees our code to be as fragile as possible w.r.t dependencies. I realize that knowing when changing dependencies break things is nice, but maybe that can be separated off somewhere else?

**Comment by jchodera** [comment]
> Is there some way to not always use the latest code? 

Yes. Step up and be involved in the OpenMM release cycle to get your bug fixes into OpenMM and releases our rapidly so I don't have to do everything.

**Comment by pgrinaway** [comment]
>Yes. Step up and be involved in the OpenMM release cycle to get your bug fixes into OpenMM and releases our rapidly so I don't have to do everything.


This is completely beside my point. I mean the idea of dependencies being able to break every night is a bad thing, not only OpenMM.

**Comment by jchodera** [comment]
This actually seems to be an issue with parmed:
```
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/parmed/openmm/topsystem.py", line 92, in load_topology
    struct.bonds.append(Bond(atommap[a1], atommap[a2]))
KeyError: <Atom 1 (CH3) of chain 0 residue 0 (ACE)>
-------------------- >> begin captured logging << --------------------
```

**Comment by pgrinaway** [comment]
>This actually seems to be an issue with parmed:

Ok, right. None of us can possibly be involved with all of the dependencies, or even cover all of them well. My point is that we don't want the builds to suddenly break because a dependency changed.

**Comment by jchodera** [comment]
> This is completely beside my point. I mean the idea of dependencies being able to break every night is a bad thing, not only OpenMM.

That's actually the philosophy behind nightly builds. When something fails due to an upstream dependency, we notice it right away and decide to either restrict ourselves to a *specific* version of a dependency, adapt to the upstream changes and update our minimum requirements, or fix the upstream issue. Otherwise, we end up with code rot, like [ensembler](http://github.com/choderalab/ensembler), which I'm *still* trying to get into working condition because it is just crazy out of date with dependencies nobody ever locked down (or dependencies that have disappeared) and nobody noticed for months.

I'm open to other ways of doing things, but if the proposal leaves open the potential for another ensembler catastrophe, I'm going to say no.

**Comment by jchodera** [comment]
> Ok, right. None of us can possibly be involved with all of the dependencies, or even cover all of them well. My point is that we don't want the builds to suddenly break because a dependency changed.

Well, now we're stuck requiring a version of OpenMM with `CustomIntegrator` bugfixes (7.1.0) and ParmEd (which requires an OpenMM prior to https://github.com/pandegroup/openmm/pull/1668). 

What are our options?

**Comment by jchodera** [comment]
> Ok, right. None of us can possibly be involved with all of the dependencies, or even cover all of them well. My point is that we don't want the builds to suddenly break because a dependency changed.

The only way we can avoid that is to pin all the dependencies to *specific* versions. That's a fine thing to do, updating them as needed to get new functionality, provided old dependencies are never deprecated.

Do you want to take a stab at this after the present situation is resolved? 

Or do you have another plan?

**Comment by jchodera** [comment]
The issue is probably not with parmed or openmm, but with perses. We can't find a way for the parmed or openmm issues to have caused this.

What's the plan now?

**Comment by jchodera** [comment]
cc: https://github.com/ParmEd/ParmEd/issues/779

**Comment by pgrinaway** [comment]
>I'm open to other ways of doing things

Cool, will defer this dicussion to later.

>The only way we can avoid that is to pin all the dependencies to specific versions. That's a fine thing to do, updating them as needed to get new functionality, provided old dependencies are never deprecated.

>Do you want to take a stab at this after the present situation is resolved?

Ok. I want to finish the Ras stuff first though.


>The issue is probably not with parmed or openmm, but with perses. We can't find a way for the parmed or openmm issues to have caused this.

So then how could the build have broken overnight with no code changes?

**Comment by jchodera** [comment]
> The only way we can avoid that is to pin all the dependencies to specific versions. That's a fine thing to do, updating them as needed to get new functionality, provided old dependencies are never deprecated.

The drawback here is that you are likely preventing your code from being installed alongside any other code that might require a conflicting version. This might work for development, though.

> So then how could the build have broken overnight with no code changes?

I'm not certain, but we can't figure out how the OpenMM change would have broken things. Is this test deterministic, or is it stochastic and could have just run into an edge case this time?

**Comment by pgrinaway** [comment]
>The drawback here is that you are likely preventing your code from being installed alongside any other code that might require a conflicting version. This might work for development, though.


Yep, agreed. I think it's fine for development.

>I'm not certain, but we can't figure out how the OpenMM change would have broken things. Is this test deterministic, or is it stochastic and could have just run into an edge case this time?

It's a deterministic failure, because it's trying to use `numpy.random.choice(a)` and `a` (the list of matches) is `None`.

**Comment by jchodera** [comment]
@pgrinaway @juliebehr : I've updated the sampler tests to restrict this to only the systems that need to pass for the first paper, but **it is essential that we remember to add to the list of test systems to test when the [`hybrid-tests`](https://github.com/choderalab/perses/tree/hybrid-tests) branch is merged**.  I'll create an issue to remind us.

**Comment by jchodera** [comment]
I've disabled the sampler tests, but we're seeing one confusing failure on a bunch of mutation topology proposal tests:
```
======================================================================
ERROR: Failure: NameError (name 'attr' is not defined)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/nose/loader.py", line 418, in loadTestsFromName
    addr.filename, addr.module)
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/perses/tests/test_samplers.py", line 84, in <module>
    @attr('advanced')
NameError: name 'attr' is not defined
======================================================================
ERROR: Make sure proposals can be made using optional argument allowed_mutations
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/perses/tests/test_topology_proposal.py", line 173, in test_specify_allowed_mutants
    pm_top_proposal = pm_top_engine.propose(system, modeller.topology)
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/perses/rjmc/topology_proposal.py", line 284, in propose
    atom_map = self._construct_atom_map(residue_map, old_topology, index_to_new_residues, new_topology)
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/perses/rjmc/topology_proposal.py", line 595, in _construct_atom_map
    _ , local_atom_map = self._get_mol_atom_matches(old_oemol_res, new_oemol_res)
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/perses/rjmc/topology_proposal.py", line 651, in _get_mol_atom_matches
    match = np.random.choice(matches)
  File "mtrand.pyx", line 1397, in mtrand.RandomState.choice (numpy/random/mtrand/mtrand.c:15477)
ValueError: a must be non-empty
======================================================================
ERROR: Propose a random mutation in insulin
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/perses/tests/test_topology_proposal.py", line 244, in test_run_point_mutation_propose
    pm_top_proposal = pm_top_engine.propose(system, modeller.topology)
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/perses/rjmc/topology_proposal.py", line 284, in propose
    atom_map = self._construct_atom_map(residue_map, old_topology, index_to_new_residues, new_topology)
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/perses/rjmc/topology_proposal.py", line 595, in _construct_atom_map
    _ , local_atom_map = self._get_mol_atom_matches(old_oemol_res, new_oemol_res)
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/perses/rjmc/topology_proposal.py", line 651, in _get_mol_atom_matches
    match = np.random.choice(matches)
  File "mtrand.pyx", line 1397, in mtrand.RandomState.choice (numpy/random/mtrand/mtrand.c:15477)
ValueError: a must be non-empty
======================================================================
ERROR: perses.tests.test_topology_proposal.test_alanine_dipeptide_map
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/perses/tests/test_topology_proposal.py", line 264, in test_alanine_dipeptide_map
    proposal = pm_top_engine.propose(system, modeller.topology)
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/perses/rjmc/topology_proposal.py", line 284, in propose
    atom_map = self._construct_atom_map(residue_map, old_topology, index_to_new_residues, new_topology)
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/perses/rjmc/topology_proposal.py", line 595, in _construct_atom_map
    _ , local_atom_map = self._get_mol_atom_matches(old_oemol_res, new_oemol_res)
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/perses/rjmc/topology_proposal.py", line 651, in _get_mol_atom_matches
    match = np.random.choice(matches)
  File "mtrand.pyx", line 1397, in mtrand.RandomState.choice (numpy/random/mtrand/mtrand.c:15477)
ValueError: a must be non-empty
======================================================================
ERROR: Make sure mutations are successful between every possible pair of before-and-after residues
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/perses/tests/test_topology_proposal.py", line 331, in test_mutate_from_every_amino_to_every_other
    pm_top_proposal = pm_top_engine.propose(current_system, current_topology)
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/perses/rjmc/topology_proposal.py", line 284, in propose
    atom_map = self._construct_atom_map(residue_map, old_topology, index_to_new_residues, new_topology)
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/perses/rjmc/topology_proposal.py", line 595, in _construct_atom_map
    _ , local_atom_map = self._get_mol_atom_matches(old_oemol_res, new_oemol_res)
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/perses/rjmc/topology_proposal.py", line 651, in _get_mol_atom_matches
    match = np.random.choice(matches)
  File "mtrand.pyx", line 1397, in mtrand.RandomState.choice (numpy/random/mtrand/mtrand.c:15477)
ValueError: a must be non-empty
======================================================================
ERROR: Test example system with certain mutations allowed to mutate
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/perses/tests/test_topology_proposal.py", line 445, in test_limiting_allowed_residues
    pl_top_proposal = pl_top_library.propose(system, modeller.topology)
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/perses/rjmc/topology_proposal.py", line 284, in propose
    atom_map = self._construct_atom_map(residue_map, old_topology, index_to_new_residues, new_topology)
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/perses/rjmc/topology_proposal.py", line 595, in _construct_atom_map
    _ , local_atom_map = self._get_mol_atom_matches(old_oemol_res, new_oemol_res)
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/perses/rjmc/topology_proposal.py", line 651, in _get_mol_atom_matches
    match = np.random.choice(matches)
  File "mtrand.pyx", line 1397, in mtrand.RandomState.choice (numpy/random/mtrand/mtrand.c:15477)
ValueError: a must be non-empty
======================================================================
ERROR: Test 'always_change' argument in topology proposal
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/perses/tests/test_topology_proposal.py", line 484, in test_always_change
    pl_top_proposal = pl_top_library.propose(system, topology)
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/perses/rjmc/topology_proposal.py", line 284, in propose
    atom_map = self._construct_atom_map(residue_map, old_topology, index_to_new_residues, new_topology)
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/perses/rjmc/topology_proposal.py", line 595, in _construct_atom_map
    _ , local_atom_map = self._get_mol_atom_matches(old_oemol_res, new_oemol_res)
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/perses/rjmc/topology_proposal.py", line 651, in _get_mol_atom_matches
    match = np.random.choice(matches)
  File "mtrand.pyx", line 1397, in mtrand.RandomState.choice (numpy/random/mtrand/mtrand.c:15477)
ValueError: a must be non-empty
======================================================================
ERROR: Test example system with peptide and library
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/perses/tests/test_topology_proposal.py", line 526, in test_run_peptide_library_engine
    pl_top_proposal = pl_top_library.propose(system, modeller.topology)
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/perses/rjmc/topology_proposal.py", line 284, in propose
    atom_map = self._construct_atom_map(residue_map, old_topology, index_to_new_residues, new_topology)
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/perses/rjmc/topology_proposal.py", line 595, in _construct_atom_map
    _ , local_atom_map = self._get_mol_atom_matches(old_oemol_res, new_oemol_res)
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/perses/rjmc/topology_proposal.py", line 651, in _get_mol_atom_matches
    match = np.random.choice(matches)
  File "mtrand.pyx", line 1397, in mtrand.RandomState.choice (numpy/random/mtrand/mtrand.c:15477)
ValueError: a must be non-empty
```
ANy idea what is going on here @pgrinaway @juliebehr? I don't think it would be wise to disable *all* mutation engine tests completely, especially when it seems like there is a single underlying cause that may be simple to fix.

File: perses/tests/test_samplers.py [hunk 1]
```diff
@@ -13,6 +13,8 @@
 # IMPORTS
 ################################################################################
 
+from nose.plugins.attrib import attr
```
Comment (jchodera) [inline perses/tests/test_samplers.py]: I'm tagging tests we don't want to run on travis with `@attr('advanced')`.

Tests labeled as `advanced` are skipped on travis via adding the `-a '!advanced'` to `nosetests` on travis.

Comment (pgrinaway) [inline perses/tests/test_samplers.py]: Oh, cool, thanks!

**Comment by pgrinaway** [comment]
> I don't think it would be wise to disable all mutation engine tests completely, especially when it seems like there is a single underlying cause that may be simple to fix.

Why not? We'd like to get this merged so that the travis tests don't burden the group, and the mutation stuff is not slated to be published for some time. I have no idea what is causing that problem, and if it is actually urgent that we make travis faster (if I were working on a different project I would agree), then I think we should have the tests only cover what is essential for the next milestone (the paper).

**Comment by jchodera** [comment]
Ok, I'll mark all the mutation tests with the same attribute. Local tests will still run to allow debugging, but travis will skip them.

**Comment by pgrinaway** [comment]
That sounds good. To be clear, I _absolutely_ want to solve these problems. I just want to prioritize:

1) Not breaking stuff we need for the paper
2) Not having our travis destroy the rest of the lab's productivity
3) Get Perses running for the paper
4) Other stuff

**Comment by jchodera** [comment]
I wanted to understand what was going on in case we broke something important in `openmm`, and I think I've tracked down the bug in the mutations. 

Currently, the mutation code in perses operates on internal private data structures for bonds that have changed due to https://github.com/pandegroup/openmm/pull/1668. We should make the public API of `Topology` much clearer and discourage operation on private internal data structures like `topology._bonds`, which is what currently happens in [`topology_proposal.py](https://github.com/choderalab/perses/blob/master/perses/rjmc/topology_proposal.py). 

I'll just comment all the mutation tests out for now. @juliebehr will have to rework the mutation code to accommodate the changes from https://github.com/pandegroup/openmm/pull/1668 later. I'll create an issue to remind us.

**Comment by jchodera** [comment]
Ready to merge @pgrinaway

**Comment by pgrinaway** [comment]
Ok, this looks good to me
